text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+        spans:                                                            # arrow type = list of struct
+          - start_time_unix_nano: timestamp                               # required 
+            end_time_unix_nano: timestamp                                 # required
+            trace_id: 16_bytes_binary | 16_bytes_binary_dictionary        # required, arrow fixed size binary array
+            span_id: 8_bytes_binary | 8_bytes_binary_dictionary           # required, arrow fixed size binary array
+            trace_state: string | string_dictionary 
+            parent_span_id: 8_bytes_binary | 8_bytes_binary_dictionary    # arrow fixed size binary array
+            name: string | string_dictionary                              # required
+            kind: uint8_dictionary                                        # OTLP enum with 6 variants 
+            attributes: *attributes
+            dropped_attributes_count: uint32 
+            events: 
+              - time_unix_nano: timestamp 
+                name: string | string_dictionary 
+                attributes: *attributes
+                dropped_attributes_count: uint32
+            dropped_events_count: uint32 
+            links: 
+              - trace_id: 16_bytes_binary | 16_bytes_binary_dictionary    # arrow fixed size binary array
+                span_id: 8_bytes_binary | 8_bytes_binary_dictionary       # arrow fixed size binary array
+                trace_state: string | string_dictionary 
+                attributes: *attributes
+                dropped_attributes_count: uint32 
+            dropped_links_count: uint32
+            status: 
+              code: uint8_dictionary                                      # OTLP enum with 4 variants 
+              status_message: string | string_dictionary
+```
+
+## Implementation Recommendations
+
+### Protocol Extension and Fallback Mechanism
+
+The support of this new protocol can only be progressive, so implementers are advised to follow the following
+implementation recommendations (phase 1):
+
+* OTLP Receiver: Listen on a single TCP port for both OTLP and OTLP Arrow. The goal is to make the support of this
+  protocol extension
+  transparent and automatic. This can be achieved by adding the `ArrowStreamService` to the same gRPC listener. A
+  configuration
+  parameter could be added to the OTLP receiver to disable this default behavior to support specific uses.
+* OTLP Exporter: By default the OTLP exporter should initiate a connection to the `ArrowStreamService` endpoint of the target
+  receiver. If this connection fails because the `ArrowStreamService` is not implemented by the target, the exporter
+  must automatically fall back on the behavior of the classic OTLP protocol. A configuration parameter could be added to
+  disable this default behavior.
+
+The implementation of these two rules should allow a seamless and adaptive integration of OTLP Arrow into the current
+ecosystem.
+
+### Batch ID Generation
+
```

Comment:
With all this ID handling, don't forget that there are levels of metadata which can be leveraged by anything processing the Arrow data. Metadata is just a list of key/value pairs, but there is optional schema level metadata and column level metadata (along with field level metadata for structs). I'm not saying it would *definitely* be applicable here, but make sure to keep in mind this metadata is available and built in to Arrow so that you could potentially leverage it.

Reply:
Yes I'm aware of this metadata layer in Arrow Records. But what are the guarantees of transferring this metadata when it goes through a processing tier like DataFusion?

Reply:
That's a fair point, I have no idea what pieces of Metadata DataFusion would maintain or not.